### PR TITLE
Use the GHC lexer for the Hyperlinker backend

### PR DESCRIPTION
**Pull request by harpocrates**
Lead up discussion is [here](https://github.com/haskell/haddock/issues/488). Moving my concerns here:

  * CPP is not handled in the GHC lexer, yet Haddock will have to handle it. Currently, the solution is to recover from _all_ lexical errors by turning problematic lines into comments (and then restarting the lexer from after that line). That seems like an OK failure mode to me, but I think it should be discussed...

 *  Language pragmas (actually _all_ top-level pragmas) are not lexed as pragmas. This (probably) requires a GHC change. In the meantime, language pragmas would look like regular comments.

    GHC's handling of top-level pragmas is a (IMHO) huge lexer-level hack to try to avoid loading the whole file and we are feeling that here. (If `bytestring` ever becomes a stage 0 package, `Data.ByteString.Lazy` is what that really needs - not a hand coded list-based parser.)

### CODE REVIEW

**Comment by harpocrates** [comment]
Alright. I think I've actually found an acceptable workaround to my two points from the previous comment. In a nutshell, we will do no better than we did before for parsing CPP and pragmas than we did before.

  - we still identify CPP based on the suboptimal criteria that the line starts with `#`
  - we still assumes anything of the form `{-# ... #-}` is a pragma

These both sound like acceptable workarounds. We could probably do better for the second point, but it would require some non-trivial GHC lexer hacking - which is probably not worth it.

TODO before this is mergeable:

  - [ ] test performance
  - [ ] update tests

**Comment by alexbiehl** [comment]
Nice! Finally a chance to throw the custom lexer over board! This sounds
reasonable! Have you ran the hyperlinker tests?

I think a lot of performance improvement in this area came with
https://github.com/haskell/haddock/commit/740458ac4d2acf197f2ef8dc94a66f9b160b9c3c
which eliminates an accidentally quietadric look-up.

Since we are working with GHCs SrcLoc and SrcSpan we can also eliminate the
custom Position and Span from haddock, right?

Alec Theriault <notifications@github.com> schrieb am Mi., 6. Dez. 2017,
06:31:

**Comment by harpocrates** [comment]
I think this is ready to merge. Sample of changes (before and after):

screen shot 2017-12-07 at 2 09 22 pm screen shot 2017-12-07 at 2 09 34 pm

Here are the only caveats I can think of:

  * Less flexibility around failure. If the lexer fails on some input, the whole chunk of Haskell source (bounded by the nearest CPP lines) will be marked as "unknown". We could conceivably just fall back onto the lexer we had before, but that isn't currently implemented. I'd like to witness _some_ failing chunk of valid Haskell before bothering to do this.
  * CPP is a hack. In a nutshell, the file gets split up into CPP chunks and non-CPP chunks. The claim is that the non-CPP chunks can be lexed just fine on their own. I'm not sure this claim _always_ holds. Besides, our heuristics for separating CPP from non-CPP could be buggy.

**Please do try this out before merging** - I'd like more confidence about performance in particular. 😄 

Thanks!

**Comment by alexbiehl** [comment]
Exciting stuff! Will try this at home!

Boy, the next haddock release is packed with new features and fixes. Awesome!

File: haddock-api/src/Haddock/Backends/Hyperlinker/Parser.hs [hunk 1]
```diff
@@ -1,214 +1,368 @@
...
+    go :: RealSrcLoc -> [Either String String] -> [Located L.Token]
+    go _   [] = []
+    go pos ls =
+      let (hLinesRight,  ls')  = span isRight ls
+          (cppLinesLeft, rest) = span isLeft ls'
+          
+          hSrc   = concat [ hLine   | Right hLine  <- hLinesRight  ]
+          cppSrc = concat [ cppLine | Left cppLine <- cppLinesLeft ]
+      
+      in case L.lexTokenStream (stringToStringBuffer hSrc) pos dflags of
```
Comment (alexbiehl) [inline haddock-api/src/Haddock/Backends/Hyperlinker/Parser.hs]: IIRC for each chunk (that is code between some CPP markers) we call the lexer. For each chunk we call `stringToStringBuffer`. If the lexer fails we make the offending part into a comment and try again with the rest, create a new `StringBuffer` and call the lexer again.  Calling `stringToStringBuffer` is pretty expensive and doing it over and over again seems unecessary.

Comment (harpocrates) [reply haddock-api/src/Haddock/Backends/Hyperlinker/Parser.hs]: I don't think that is happening. If the lexer fails, that chunk gets marked as one big "unknown" token and we move on (see the `PFailed` case - we just shove `hSrc` into one big token).

**Comment by alexbiehl** [comment]
Right! The result looks really good!

**Comment by harpocrates** [comment]
Thanks for reviewing - I really appreciate how responsive you've been!

**Pull request merged**